Mathlib/RingTheory/PowerSeries/WeierstrassPreparation.lean

```diff
@@ -0,0 +1,617 @@
+
+In this file we define Weierstrass division, Weierstrass factorization, and prove
+Weierstrass preparation theorem.
+
+We assume that a ring is adic complete with respect to some ideal.
+If such ideal is a maximal ideal, then by `isLocalRing_of_isAdicComplete_maximal`,
+such ring has only on maximal ideal, and hence such ring is a complete local ring.
+
+## Main definitions
+
+- `PowerSeries.IsWeierstrassDivisionAt f g q r I`: a `Prop` which asserts that a power series
+  `q` and a polynomial `r` of degree `< n` satisfy `f = g * q + r`, where `n` is the order of the
+  image of `g` in `(A / I)⟦X⟧` (defined to be zero if such image is zero, in which case
+  it's mathematically not considered).
+
+- `PowerSeries.IsWeierstrassDivision`: version of `PowerSeries.IsWeierstrassDivisionAt`
+  for local rings with respect to its maximal ideal.
+
+- `PowerSeries.IsWeierstrassFactorizationAt g f h I`: a `Prop` which asserts that `f` is a
+  distingushed polynomial at `I`, `h` is a formal power series over `A` which a unit, such that
+  the formal power series `g` satisfies `g = f * h`.
+
+- `PowerSeries.IsWeierstrassFactorization`: version of `PowerSeries.IsWeierstrassFactorizationAt`
+  for local rings with respect to its maximal ideal.
+
+## Main results
+
+- `PowerSeries.exists_isWeierstrassDivision`: **Weierstrass division**
+  ([washington_cyclotomic], Proposition 7.2): let `f`, `g` be power series
+  over a complete local ring, such that the image of `g` in the residue field is not zero.
+  Let `n` be the order of the image of `g` in the residue field. Then there exists a power series
+  `q` and a polynomial `r` of degree `< n`, such that `f = g * q + r`.
+
+- `PowerSeries.IsWeierstrassDivision.elim`: The `q` and `r` in the Weierstrass division is unique.
+
+- `PowerSeries.exists_isWeierstrassFactorization`: **Weierstrass preparation theorem**
+  ([washington_cyclotomic], Theorem 7.3): let `g` be a power series
+  over a complete local ring, such that the image of `g` in the residue field is
+  not zero. Then there exists a distinguished polynomial `f` and a power series `h`
+  which is a unit, such that `g = f * h`.
+
+- `PowerSeries.IsWeierstrassFactorization.elim`: The `f` and `h` in Werierstrass preparation
+  theorem is unique.
+
+## References
+
+- [Washington, Lawrence C. *Introduction to cyclotomic fields.*][washington_cyclotomic]
+
+-/
+
+open scoped Polynomial PowerSeries
+
+namespace PowerSeries
+
+variable {A : Type*} [CommRing A]
+
+/-!
+
+## Weierstrass division
+
+-/
+
+section IsWeierstrassDivisionAt
+
+variable (f g q : A⟦X⟧) (r : A[X]) (I : Ideal A)
+
+/-- `PowerSeries.IsWeierstrassDivisionAt f g q r I` is a `Prop` which asserts that a power series
+`q` and a polynomial `r` of degree `< n` satisfy `f = g * q + r`, where `n` is the order of the
+image of `g` in `(A / I)⟦X⟧` (defined to be zero if such image is zero, in which case
+it's mathematically not considered). -/
+def IsWeierstrassDivisionAt : Prop :=
```

Comment:
I think it is a good idea to make it a structure instead of using `∧`, but as you prefer (you can use `@[mk_iff]` to make the relevant iff result).

Reply:
Sure.